shared/src/main/scala/pl/metastack/metarouter/Router.scala

```diff
@@ -0,0 +1,148 @@
+                                            map: FlatMapper.Aux[Args.Convert.type, ROUTE, L]) = {
+    val f: String => Option[Any] = Router.parse(other, _)
+    new Router(parsers :+ f)
+  }
+
+  def parse(uri: String): Option[Any] =
+    parsers.foldLeft(Option.empty[Any]) { case (acc, cur) =>
+      acc.orElse(cur(uri))
+    }
+}
+
+object Router {
+  private[metarouter] class MappedRouterHelper[T] {
+    def apply[ROUTE <: HList, Params <: HList](route: Route[ROUTE])
+                                              (implicit p: Generic.Aux[T, Params],
+                                                       ev: FlatMapper.Aux[Args.Convert.type, ROUTE, Params]
+                                              ): MappedRoute[ROUTE, T] =
+      new MappedRoute[ROUTE, T](route)
+  }
+
+  def route[T] = new MappedRouterHelper[T]
+
+  private[metarouter] def split(s: String): List[String] = {
+    val x = s.stripPrefix("/")
+    if (x.isEmpty) Nil
+    else x.split('/').toList
+  }
+
+  def create[ROUTE <: HList, T, L <: HList](route: MappedRoute[ROUTE, T])
+                                           (implicit gen: Generic.Aux[T, L],
+                                                     map: FlatMapper.Aux[Args.Convert.type, ROUTE, L]) =
+    new Router(List(Router.parse(route, _)))
+
+  // TODO Figure out what to do with relative routes and query parameters
+  def parse(s: String): RouteData[HList, HNil] = {
+    val r = Route(split(s).foldRight[HList](HNil)((x, y) => x :: y))
+    RouteData(r, HNil)
+  }
+
+  def parse[ROUTE <: HList, L <: HList](route: Route[ROUTE], s: String)
+    (implicit map: FlatMapper.Aux[Args.Convert.type, ROUTE, L]):
+    Option[RouteData[ROUTE, L]] =
+  {
+    import shapeless.HList.ListCompat._
+
+    def m[R <: HList](r: R, s: Seq[String]): Option[HList] =
+      (r, s) match {
+        case (HNil, Nil) => Some(HNil)
+        case (_,    Nil) => None
+        case (HNil, _)   => None
+        case ((rH: String)  #: rT, sH :: sT) if rH == sH => m(rT, sT)
+        case ((rH: String)  #: rT, sH :: sT) => None
+        case ((arg: Arg[_]) #: rT, sH :: sT) =>
+          arg.parseableArg.urlDecode(sH).flatMap { decoded =>
+            m(rT, sT).map(decoded :: _)
+          }
+      }
+
+    m[ROUTE](route.pathElements, split(s)).map { x =>
+      // Using asInstanceOf as a band aid since compiler isn't able to confirm the type.
+      RouteData(route, x.asInstanceOf[L])
+    }
+  }
+
+  def parse[ROUTE <: HList, T, Args <: HList](mappedRoute: MappedRoute[ROUTE, T], uri: String)
+                                             (implicit gen: Generic.Aux[T, Args],
+                                                       map: FlatMapper.Aux[Args.Convert.type, ROUTE, Args]): Option[T] =
+    parse(mappedRoute.route, uri).map(parsed => gen.from(parsed.data))
+
+  /**
+    * Converts all the chunks of the path to `HR` using the passed `~>>` function.
+    * Then combines all the `HR` elements together.
+    */
+  def fold[
```

Comment:
This method itself was implemented because I did not know about `foldMap` at the point. `foldMap` does almost exactly what this one does, so we should stick to the standard library rather than reinventing the wheel.